### DISCUSSION TITLE
M5StickCPlus fail ledc function

### DISCUSSION THREAD

**Discussion by kagjerde** [PlummersSoftwareLLC/NightDriverStrip#238] [Q&A]
I have built and uploaded several examples for the M5 Stick C Plus without any errors, but no output on any pins or LCD. 
The ESP re-boots after the LEDC fail to initialize. 

I have tried to upgrade the toolchain, but still the same error.

This is the error in the serial monitor in bold:

> (I) (PrintOutputHeader)(C1) ------------------------------------------------------------------------------------------------------------
> (I) (PrintOutputHeader)(C1) M5STICKC: 0, USE_M5DISPLAY: 1, USE_OLED: 0, USE_TFTSPI: 0, USE_LCD: 0, USE_AUDIO: 1, ENABLE_REMOTE: 1
> (I) (PrintOutputHeader)(C1) Version 36: Wifi SSID: mywifi - ESP32 Free Memory: 249364, PSRAM:0, PSRAM Free: 0
> (I) (PrintOutputHeader)(C1) ESP32 Clock Freq : 240 MHz
> (I) (setup)(C1) Startup!
> (I) (setup)(C1) Starting DebugLoopTaskEntry
> (W) (StartDebugThread)(C1) >> Launching Debug Thread
> (I) (DebugLoopTaskEntry)(C0) >> DebugLoopTaskEntry
> (I)
> (I) (setup)(C1) Intializizing TFT display
> (I)
> M5StickCPlus initializing...OK
> E (1250) ledc: freq_hz=0 duty_resolution=13
> **[  1267][E][esp32-hal-ledc.c:75] ledcSetup(): ledc setup failed!
> E (1253) ledc: ledc_get_duty(739): LEDC is not initialized**
> ets Jun  8 2016 00:22:57
> 
> rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
> configsip: 188777542, SPIWP:0xee
> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
> mode:DIO, clock div:2
> load:0x3fff0030,len:1184
> load:0x40078000,len:13192
> load:0x40080400,len:3028
> entry 0x400805e4
> E (694) esp_core_dump_flash: No core dum���ѥѥ���found!
> E (694) esp_core_dump_flash: No core dumReplacing Idle Tasks with TaskManager...
> (I) (PrintOutputHeader)(C1) NightDriverStrip

Any suggestions?

**Comment by davepl** [comment]
I think the M5STICKC: 0 is telling - that should be defined.

Build the SPECTRUM project and upload that.  You should get a sound analyzer on the display even with nothing else connected.  Start there and let me know what happens!

Cheers,
Dave

**Reply by TWM617** [reply]
Since I am seeing the exact same problem, I loaded the SPECTRUM project and it works perfectly.

Regards, Tim.

**Comment by TWM617** [comment]
LEDC failing may be the problem for laserline but m5plusdemo also reports that LEDC fails but seems to run fine.

**Comment by rbergen** [comment]
I noticed the logging mentions `ledcSetup()`. In the current source tree, that function is only called for the Umbrella project (platformio.ini: `env:umbrella`). That project is built for the Espressif ESP32 Dev Kit boards (platformio.ini: `board = esp32dev`), not the M5StickC Plus.

**Comment by TWM617** [comment]
Agree. Hard to understand why a laserline project would generate an ledc error on an M5StickC Plus.

Tim
________________________________

**Comment by davepl** [comment]
Any chance it’s an ESP32-S3?  There are some differences in what the chips support, like DMA, and the S3 only supports “slow mode” for ledc.

**Comment by TWM617** [comment]
No, it is in an orange case with M5 embossed into the switch and M5StickC Plus 1.1 printed on the back.

Tim
________________________________

**Comment by kagjerde** [comment]
I have tested a bit further and found that if I use the M5Stick-C (not the plus version) code (like M5demo or spectrumlite) - the led output and webserver works.
The same if I define the M5Scick_C in Laserline, the led output also works.
But the display does not work at all for any examples.

M5Stick C uses ST7735 chip set and the plus version ST7789v2 (actually ST7789 is used in the M5StickCPlus library..), so for the display to work we need the plus library.

For all builds with M5Stick_c_plus, the chip boots OK until it reboots:

> M5StickCPlus initializing...OK
> E (1250) ledc: freq_hz=0 duty_resolution=13
> [ 1267][E][esp32-hal-ledc.c:75] ledcSetup(): ledc setup failed!
> E (1253) ledc: ledc_get_duty(739): LEDC is not initialized

I tried to comment out the ledc functions, but the chip then reboots right after initializing...OK, so I do not think the ledc function causes the boot (ledc is used in the M5StickCPlus library for speaker and M5display).

Changing the definition in spectrum from plus version to old MSStick_c does not generate any errors and the led output works - but the display still does not work due to incompatible chipset. 

I believe the error is related to the M5 plus library, since all other libraries remain the same.
I also get the same error when upgrading to the latest 0.0.8 version.. 

I will test some more unless someone have any ideas

**Reply by TWM617** [reply]
Nice work! What baffles me is that spectrum works on the plus display and apparently the LEDs work on the non-plus but both don't work together. I would have to guess spectrum uses the plus library but not the same one as the non-plus?

**Comment by davepl** [comment]
I presume the headers account for the differing chipsets:  so if you define M5STICK, you get the header for the base version.  If you define plus, you get the header for the plus version:

#if M5STICKC
#define LED_BUILTIN 10                          // Not defined by the M5 headers, but it seems to be PIN 10
#include "M5StickC.h"
#undef min                                      // They define a min() on us
#endif

#if M5STICKCPLUS
#define LED_BUILTIN 10                          // Not defined by the M5 headers, but it seems to be PIN 10
#include "M5StickCPlus.h"
#undef min                                      // They define a min() on us
#endif

Can you make sure that M5STICKCPLUS isn’t defined for your non-plus case?

I can sure believe its broken at some point, as I don’t have a non-plus version anymore, but wrote it on the non-plus originally and never added chipset support for the new one, so it’s almost got to be a header issue!

Appreciate your help tracking this one down!

- Dave

**Comment by kagjerde** [comment]
In Laserline I have built with dev_m5stick-c-plus and dev_m5stick-c, and searched in .pio/libdeps if plus or non-plus definition found in either.

The result is that everything seems fine, no traces of plus version in m5stick-c,, and no non-plus version in m5stick-c-plus.

In the M5Stock hardware the m5stick-c works fine - but no display (i assume it has no display anyway)
For m5stick-c-plus build, it re-boots all the time:

> E (488) esp_core_dump_flash: No core duy�partition found!
> E (488) esp_core_dump_flash: No core dump partition found!
> Replacing Idle Tasks with TaskManager...
> (I) (setup)(C1) Intializizing TFT display
> (I) 
> M5StickCPlus initializing...OK
> E (1212) ledc: freq_hz=0 duty_resolution=13
> [  1227][E][esp32-hal-ledc.c:75] ledcSetup(): ledc setup failed!
> E (1214) ledc: ledc_get_duty(739): LEDC is not initialized
> ets Jun  8 2016 00:22:57
> 
> rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
> configsip: 188777542, SPIWP:0xee
> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
> mode:DIO, clock div:2
> load:0x3fff0030,len:1184
> load:0x40078000,len:13192
> load:0x40080400,len:3028
> entry 0x400805e4

Does anybody else have the same problem with the M5Stick plus v.1.1 re-booting ? 
If I am the only one who experience this problem it may be due to a faulty unit or some other fundamental error.

I have also tried to define the new display chip ST7789 normal and V2 version in the m5stick-c environment, but so far i have not been able to bring the plus version display to life. Next I plan to use the M5Stick-C_plus library to test the function.
I have however used the M5 Stick plus factory test to successfully erase and burn the demo software. 
This software can be found here [M5 test](https://docs.m5stack.com/en/core/m5stickc_plus), so at least it seems to be working with the demo software with the screen enabled.

Thanks again for all your support.

**Comment by TWM617** [comment]
I have exactly the same problem.

Tim
________________________________

**Comment by TWM617** [comment]
OK, I figured it out. The only problem is that the LED PIN0 was not set in laserline as it was in the other patterns so it defaulted to 5 and in my case, it should have been 32. When I defined it in laserline, voila!

**Reply by kagjerde** [reply]
The above does not fix the re-booting problem on the M5stickcPlus version, but downloading from the WEB installer did work also with the display. 

After some testing I found that I have an old TFT_eSPI library installed as a separate installation used for other ESP projects. 
This caused the compiler to not use the TFT_eSPI library defined under M5StickCPlus @ 0.0.2.

After deleting the independent TFT_eSPI library all works as intended.

**Reply by TWM617** [reply]
Apparently my problem was not the same though it exhibited similar symptoms. Glad that we are both working.
